File: client/lib/medicalVault.ts
import { scrypt } from "@noble/hashes/scrypt";
import type { VaultAead, VaultKdfParams } from "@shared/api";

const KEY_DERIVATION_SALT_BYTES = 16;
const KEY_DERIVATION_IV_BYTES = 12;
const DEK_BYTES = 32;
const DEFAULT_AEAD: VaultAead = "aes-256-gcm";

const DEFAULT_KDF_PARAMS: VaultKdfParams = {
  algo: "scrypt",
  N: 2 ** 15,
  r: 8,
  p: 1,
  keyLen: 32,
};

export const arrayBufferToBase64 = (buffer: ArrayBuffer) => {
  const bytes = new Uint8Array(buffer);
  let binary = "";
  bytes.forEach((byte) => {
    binary += String.fromCharCode(byte);
  });
  if (typeof globalThis.btoa === "function") {
    return globalThis.btoa(binary);
  }
  return Buffer.from(binary, "binary").toString("base64");
};

export const base64ToArrayBuffer = (base64: string) => {
  const binary = typeof globalThis.atob === "function"
    ? globalThis.atob(base64)
    : Buffer.from(base64, "base64").toString("binary");
  const bytes = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }
  return bytes.buffer;
};

export const deriveKey = async (secret: string, salt: Uint8Array, params: VaultKdfParams) => {
  if (params.algo === "argon2id") {
    // Argon2id preferred, but we fall back to scrypt in-browser when Argon2 WASM isn't available.
    const keyBytes = scrypt(secret, salt, {
      N: 2 ** 15,
      r: 8,
      p: 1,
      dkLen: params.keyLen,
    });
    return globalThis.crypto.subtle.importKey("raw", keyBytes, "AES-GCM", false, ["encrypt", "decrypt"]);
  }

  if (params.algo === "scrypt") {
    const keyBytes = scrypt(secret, salt, {
      N: params.N,
      r: params.r,
      p: params.p,
      dkLen: params.keyLen,
    });
    return globalThis.crypto.subtle.importKey("raw", keyBytes, "AES-GCM", false, ["encrypt", "decrypt"]);
  }

  // PBKDF2 fallback if scrypt is unavailable.
  const baseKey = await globalThis.crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(secret),
    "PBKDF2",
    false,
    ["deriveKey"],
  );
  return globalThis.crypto.subtle.deriveKey(
    { name: "PBKDF2", salt, iterations: params.iterations, hash: params.hash },
    baseKey,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"],
  );
};

export type WrappedDek = {
  wrappedKey: string;
  kdfSalt: string;
  kdfParams: VaultKdfParams;
  wrapIv: string;
  aead: VaultAead;
};

export const wrapDEK = async (
  dek: CryptoKey,
  secret: string,
  kdfParams: VaultKdfParams = DEFAULT_KDF_PARAMS,
): Promise<WrappedDek> => {
  // Wrap the random DEK with a KEK derived from the user secret (password or recovery key).
  const rawKey = await globalThis.crypto.subtle.exportKey("raw", dek);
  const salt = globalThis.crypto.getRandomValues(new Uint8Array(KEY_DERIVATION_SALT_BYTES));
  const iv = globalThis.crypto.getRandomValues(new Uint8Array(KEY_DERIVATION_IV_BYTES));
  const kek = await deriveKey(secret, salt, kdfParams);
  const wrapped = await globalThis.crypto.subtle.encrypt({ name: "AES-GCM", iv }, kek, rawKey);
  return {
    wrappedKey: arrayBufferToBase64(wrapped),
    kdfSalt: arrayBufferToBase64(salt.buffer),
    kdfParams,
    wrapIv: arrayBufferToBase64(iv.buffer),
    aead: DEFAULT_AEAD,
  };
};

export const unwrapDEK = async (payload: WrappedDek, secret: string) => {
  const kek = await deriveKey(secret, new Uint8Array(base64ToArrayBuffer(payload.kdfSalt)), payload.kdfParams);
  const decrypted = await globalThis.crypto.subtle.decrypt(
    { name: "AES-GCM", iv: new Uint8Array(base64ToArrayBuffer(payload.wrapIv)) },
    kek,
    base64ToArrayBuffer(payload.wrappedKey),
  );
  return globalThis.crypto.subtle.importKey("raw", decrypted, "AES-GCM", true, ["encrypt", "decrypt"]);
};

export const encryptDoc = async (dek: CryptoKey, data: ArrayBuffer, aad?: string) => {
  const iv = globalThis.crypto.getRandomValues(new Uint8Array(KEY_DERIVATION_IV_BYTES));
  const additionalData = aad ? new TextEncoder().encode(aad) : undefined;
  const ciphertext = await globalThis.crypto.subtle.encrypt(
    { name: "AES-GCM", iv, additionalData },
    dek,
    data,
  );
  return {
    iv: arrayBufferToBase64(iv.buffer),
    ciphertext: arrayBufferToBase64(ciphertext),
    aad,
  };
};

export const decryptDoc = async (dek: CryptoKey, payload: { iv: string; ciphertext: string; aad?: string }) => {
  const iv = new Uint8Array(base64ToArrayBuffer(payload.iv));
  const additionalData = payload.aad ? new TextEncoder().encode(payload.aad) : undefined;
  return globalThis.crypto.subtle.decrypt(
    { name: "AES-GCM", iv, additionalData },
    dek,
    base64ToArrayBuffer(payload.ciphertext),
  );
};

export const generateRecoveryKey = () => {
  const bytes = globalThis.crypto.getRandomValues(new Uint8Array(DEK_BYTES));
  const base64 = globalThis.btoa(String.fromCharCode(...bytes));
  return base64.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
};

export const getKeyStorageKey = (email?: string) =>
  `docnearme_medical_records_dek_${email ? email.toLowerCase() : "unknown"}`;

export const getStoredVaultKey = async (email?: string) => {
  const storedKey = localStorage.getItem(getKeyStorageKey(email));
  if (!storedKey) return null;
  const rawKey = base64ToArrayBuffer(storedKey);
  return globalThis.crypto.subtle.importKey("raw", rawKey, "AES-GCM", true, ["encrypt", "decrypt"]);
};

export const storeLocalVaultKey = async (email: string | undefined, key: CryptoKey) => {
  const rawKey = await globalThis.crypto.subtle.exportKey("raw", key);
  localStorage.setItem(getKeyStorageKey(email), arrayBufferToBase64(rawKey));
};

export const clearLocalVaultKey = (email?: string) => {
  localStorage.removeItem(getKeyStorageKey(email));
};

File: ios/DocNearMeiOS/Sources/DocNearMeiOS/MedicalVault.swift
import CryptoKit
import Foundation
import CryptoSwift

enum MedicalVault {
    static let keyDerivationSaltBytes = 16
    static let keyDerivationIvBytes = 12
    static let dekBytes = 32
    static let defaultAead = "aes-256-gcm"

    static let defaultKdfParams = VaultKdfParams(
        algo: "scrypt",
        N: 1 << 15,
        r: 8,
        p: 1,
        keyLen: 32,
        iterations: nil,
        hash: nil,
        opslimit: nil,
        memlimit: nil
    )

    struct WrappedDek {
        let wrappedKey: String
        let kdfSalt: String
        let kdfParams: VaultKdfParams
        let wrapIv: String
        let aead: String
    }

    static func generateRecoveryKey() -> String {
        var bytes = [UInt8](repeating: 0, count: dekBytes)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        let data = Data(bytes)
        return base64UrlEncode(data)
    }

    static func storeLocalVaultKey(email: String?, key: SymmetricKey) {
        let keyData = key.withUnsafeBytes { Data($0) }
        let account = keyStorageKey(email: email)
        _ = KeychainStore.save(keyData, account: account)
    }

    static func getStoredVaultKey(email: String?) -> SymmetricKey? {
        let account = keyStorageKey(email: email)
        guard let data = KeychainStore.load(account: account) else { return nil }
        return SymmetricKey(data: data)
    }

    static func clearLocalVaultKey(email: String?) {
        let account = keyStorageKey(email: email)
        KeychainStore.delete(account: account)
    }

    static func keyStorageKey(email: String?) -> String {
        let normalized = (email ?? "unknown").lowercased()
        return "docnearme_medical_records_dek_\(normalized)"
    }

    static func wrapDEK(_ dek: SymmetricKey, secret: String, kdfParams: VaultKdfParams = defaultKdfParams) throws -> WrappedDek {
        let salt = randomBytes(count: keyDerivationSaltBytes)
        let iv = randomBytes(count: keyDerivationIvBytes)
        let kek = try deriveKey(secret: secret, salt: salt, params: kdfParams)
        let rawKey = dek.withUnsafeBytes { Data($0) }
        let sealed = try AES.GCM.seal(rawKey, using: kek, nonce: AES.GCM.Nonce(data: iv))
        let ciphertext = sealed.ciphertext + sealed.tag

        return WrappedDek(
            wrappedKey: base64Encode(ciphertext),
            kdfSalt: base64Encode(salt),
            kdfParams: kdfParams,
            wrapIv: base64Encode(iv),
            aead: defaultAead
        )
    }

    static func unwrapDEK(payload: WrappedDek, secret: String) throws -> SymmetricKey {
        let salt = base64Decode(payload.kdfSalt)
        let iv = base64Decode(payload.wrapIv)
        let kek = try deriveKey(secret: secret, salt: salt, params: payload.kdfParams)
        let data = base64Decode(payload.wrappedKey)
        guard data.count > 16 else {
            throw NSError(domain: "DocNearMe", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid wrapped key data."])
        }
        let cipher = data.prefix(data.count - 16)
        let tag = data.suffix(16)
        let sealed = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: iv), ciphertext: cipher, tag: tag)
        let rawKey = try AES.GCM.open(sealed, using: kek)
        return SymmetricKey(data: rawKey)
    }

    static func encryptDoc(_ dek: SymmetricKey, data: Data, aad: String?) throws -> (iv: String, ciphertext: String, aad: String?) {
        let iv = randomBytes(count: keyDerivationIvBytes)
        let aadData = aad?.data(using: .utf8)
        let sealed = try AES.GCM.seal(data, using: dek, nonce: AES.GCM.Nonce(data: iv), authenticating: aadData ?? Data())
        let ciphertext = sealed.ciphertext + sealed.tag
        return (iv: base64Encode(iv), ciphertext: base64Encode(ciphertext), aad: aad)
    }

    static func decryptDoc(_ dek: SymmetricKey, payload: VaultDocDetail) throws -> Data {
        let iv = base64Decode(payload.iv)
        let data = base64Decode(payload.ciphertext)
        guard data.count > 16 else {
            throw NSError(domain: "DocNearMe", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid encrypted payload."])
        }
        let cipher = data.prefix(data.count - 16)
        let tag = data.suffix(16)
        let aadData = payload.aad?.data(using: .utf8) ?? Data()
        let sealed = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: iv), ciphertext: cipher, tag: tag)
        return try AES.GCM.open(sealed, using: dek, authenticating: aadData)
    }

    private static func deriveKey(secret: String, salt: Data, params: VaultKdfParams) throws -> SymmetricKey {
        let password = Array(secret.utf8)
        let saltBytes = Array(salt)

        switch params.algo {
        case "scrypt":
            guard let N = params.N, let r = params.r, let p = params.p, let keyLen = params.keyLen else {
                throw NSError(domain: "DocNearMe", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid scrypt parameters."])
            }
            let keyBytes = try Scrypt(password: password, salt: saltBytes, dkLen: keyLen, N: N, r: r, p: p).calculate()
            return SymmetricKey(data: Data(keyBytes))
        case "pbkdf2":
            guard let iterations = params.iterations, let keyLen = params.keyLen else {
                throw NSError(domain: "DocNearMe", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid PBKDF2 parameters."])
            }
            let keyBytes = try PKCS5.PBKDF2(
                password: password,
                salt: saltBytes,
                iterations: iterations,
                keyLength: keyLen,
                variant: .sha2(.sha256)
            ).calculate()
            return SymmetricKey(data: Data(keyBytes))
        default:
            throw NSError(domain: "DocNearMe", code: 0, userInfo: [NSLocalizedDescriptionKey: "Unsupported KDF on iOS."])
        }
    }

    private static func randomBytes(count: Int) -> Data {
        var bytes = [UInt8](repeating: 0, count: count)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes)
    }

    private static func base64Encode(_ data: Data) -> String {
        data.base64EncodedString()
    }

    private static func base64Decode(_ value: String) -> Data {
        Data(base64Encoded: value) ?? Data()
    }

    private static func base64UrlEncode(_ data: Data) -> String {
        let base64 = data.base64EncodedString()
        return base64
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
